lib/api-helper/project/support/cacheKey.ts

```diff
@@ -14,14 +14,32 @@
  * limitations under the License.
  */
 
-import { RemoteRepoRef } from "@atomist/automation-client/lib/operations/common/RepoId";
 import { sprintf } from "sprintf-js";
+import { ProjectLoadingParameters } from "../../../spi/project/ProjectLoader";
 
 /**
- * Compute a cache key from the given remote repo ref and sha
+ * Compute a cache key from the given ProjectLoadingParameters.
+ *
+ * In this context undefined and false are functionally equivalent, so
+ * undefined values are coerced to false.
+ *
  * @param {RemoteRepoRef} id
  * @return {any}
  */
-export function cacheKeyForSha(id: RemoteRepoRef): string {
-    return sprintf("%s:%s:%s:%s@%s", id.owner, id.repo, id.branch, id.sha, id.url);
+export function cacheKey(params: ProjectLoadingParameters): string {
+    return sprintf("%s:%s:%s:%s:%s:%s:%s:%s:%s@%s",
+        params.id.owner,
+        params.id.repo,
+        params.id.branch,
+        params.id.sha,
+        coerceUndefined(params?.cloneOptions?.keep),
+        coerceUndefined(params?.cloneOptions?.alwaysDeep),
+        coerceUndefined(params?.cloneOptions?.noSingleBranch),
+        coerceUndefined(params?.cloneOptions?.depth),
+        coerceUndefined(params?.cloneOptions?.detachHead),
```

Comment:
This approach would work for the boolean properties, but not for depth. Perhaps something more direct would be better:

```typescript
params.cloneOptions?.keep || false,
params.cloneOptions?.alwaysDeep || false,
params.cloneOptions?.noSingleBranch || false,
params.cloneOptions?.depth || 1,
params.cloneOptions?.detachHead || false,
```

With defaults derived from the code

-   https://github.com/atomist/automation-client/blob/master/lib/project/git/GitCommandGitProject.ts#L355
-   https://github.com/atomist/automation-client/blob/5b43b968067206df081791d3cca6456eeebeeadf/lib/spi/clone/DirectoryManager.ts#L40-L43

Reply:
works for me, will update and revert.

Sanity check - It seems like we can assume in this context that cloneOptions will always be defined?

Reply:
If it isn't, the `?.` will handle it.